sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/V2Aggregator.scala

```diff
@@ -0,0 +1,70 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.expressions.aggregate
+
+import java.io.{ByteArrayInputStream, ByteArrayOutputStream, ObjectInputStream, ObjectOutputStream}
+
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.{Expression, UnsafeProjection}
+import org.apache.spark.sql.connector.catalog.functions.{AggregateFunction => V2AggregateFunction}
+import org.apache.spark.sql.types.DataType
+
+case class V2Aggregator[BUF <: java.io.Serializable, OUT](
+    aggrFunc: V2AggregateFunction[BUF, OUT],
+    children: Seq[Expression],
+    mutableAggBufferOffset: Int = 0,
+    inputAggBufferOffset: Int = 0) extends TypedImperativeAggregate[BUF] {
+  private[this] lazy val inputProjection = UnsafeProjection.create(children)
+
+  override def nullable: Boolean = aggrFunc.isResultNullable
+  override def dataType: DataType = aggrFunc.resultType()
+  override def createAggregationBuffer(): BUF = aggrFunc.newAggregationState()
+
+  override def update(buffer: BUF, input: InternalRow): BUF = {
```

Comment:
For `TypedImperativeAggregate`, the entire `update` function is a black box to Catalyst. The magic method can't help here as codegen doesn't apply here.

We need to have deeper codegen integration in  `TypedImperativeAggregate`, which we may not able to complete before 3.2. I think it's OK, and we can always add magic method later as it won't break backward compatibility.

@sunchao can you revisit the doc of v2 `AggregateFunction` and remove descriptions of the magic method?

Reply:
Yes I think it may require some work to fully support codegen and magic method in aggregate function. How about we change the `AggregateFunction#update` to non-default method so users have to implement it? We can switch to default and throw exception when we have the magic method support in future and it should be backward compatible. cc @rdblue